fix: cross-platform prebuild downloads

mmaietta opened the pull request:
Allow prebuilt binaries to be downloaded for target platform instead of hardcoded `process.platform`. Throws an error when attempting to cross-compile (`buildFromSource` || no prebuilds found)
Example: https://github.com/electron/rebuild/blob/cb372cdf09d9ca3ac8f1231608e4ebedddabe51c/src/module-type/node-pre-gyp.ts#L31

Origin of issue: https://github.com/electron/rebuild/blob/cb372cdf09d9ca3ac8f1231608e4ebedddabe51c/src/rebuild.ts#L135

Updated interface with new `platform` option and added docs

mmaietta commented on src/module-rebuilder.ts:
```diff
@@ -151,6 +151,10 @@ export class ModuleRebuilder {
 
+    if (this.rebuilder.platform !== process.platform) {
+      throw new Error("OS Limitation - It is not possible to cross-compile native modules from source")
+    }
```
I took a look at that approach, but the reason I didn't was because then it seemed like electron-rebuild would try to rebuild every native module and throw an individual error each time
https://github.com/electron/rebuild/blob/cb372cdf09d9ca3ac8f1231608e4ebedddabe51c/src/rebuild.ts#L237-L241

Instead, I felt that putting the check at the beginning would be an early exit from the flow

mmaietta replied:
Following up here, I ended up moving the logic into `rebuildModule` as you suggested to simplify/isolate behaviors for unit testing specific rebuild/prebuild files as opposed to pulling the entire `ModuleRebuilder` into `module-type-node-gyp.ts` test case

erickzhao commented on src/rebuild.ts:
```diff
@@ -19,6 +19,13 @@ export interface RebuildOptions {
   electronVersion: string;
+  /**
+   * Override the target platform to something other than the host system platform.
+   * Note: This only applies to downloading prebuilt binaries. **It is not possible to cross-compile native modules**
```
```suggestion
   * Note: This only applies to downloading prebuilt binaries. **It is not possible to cross-compile native modules.**
```

Tiny punctuation nit but otherwise LGTM

mmaietta commented:
Writing a few tests for the code coverage requirement. Stay tuned.

mmaietta commented:
Good thing I added tests. I had missed adding `platform` to the cache key.

**I could use a tad bit of help though.** (I'll continue investigating as well) I'm running into this test failure for the "early exit" when attempting cross-compilation via node-gyp.
```
  1 failing

  1) node-gyp
       buildArgs
         cross-compilation
           throws error early:
     AssertionError: expected [Function executor] to throw an error
```
Oddly enough, if I `console.log` right before the `throw new Error`, it logs the state correctly. So for some reason, `expect(executor).throws()` is not receiving the `Error` being thrown. Any idea why?
https://github.com/electron/rebuild/blob/fff22551593b4f41a51d8420164a4233d0b47c44/src/module-type/node-gyp/node-gyp.ts#L86-L89
https://github.com/electron/rebuild/blob/fff22551593b4f41a51d8420164a4233d0b47c44/test/module-type-node-gyp.ts#L69-L74

Also, quick note, I changed `platform: string` to `platform: NodeJS.Platform` which shouldn't change much for the end user since `platform` was never configurable beforehand, so it just helps intellisense out for end users now.

mmaietta commented:
Investigation update. It's throwing the error as written, but the event isn't propagating to mocha specifically. It properly logs the error being thrown if I add a `.catch(console.error)` to `nodeGyp.rebuildModule()`. I've taken a different approach in using an error message equality (to make sure we're testing this specific error being thrown) and then `expect` on a boolean flag

All tests are passing 🙃 

https://github.com/electron/rebuild/blob/0fbaba08cc7055f394a3ad231ee630d2cbb6b663/test/module-type-node-gyp.ts#L75-L82

MarshallOfSound commented:
@mmaietta `throws` is used for synchronous checks, because the function you're calling is async you need to use the `eventually` assertion variants.

E.g. `.to.eventually.be.rejectedWith` (there are examples of this in other tests)

mmaietta commented:
Thanks @MarshallOfSound! I had overlooked the `chaiAsPromised`. Updated the test accordingly

MarshallOfSound commented on src/rebuild.ts:
```diff
@@ -151,6 +158,7 @@ export class Rebuilder implements IRebuilder {
     this.lifecycle = options.lifecycle;
     this.buildPath = options.buildPath;
     this.electronVersion = options.electronVersion;
+    this.platform = options.platform || process.platform
```
Missing a semicolon, surprised lint didn't complain

mmaietta replied:
👍 
Double-checked all new lines for semicolons and added where I had missed them. PR updated.

mmaietta replied:
@MarshallOfSound I've opened a separate PR for fixing the semicolon lint issue 🙂 
https://github.com/electron/rebuild/pull/1154

samuelmaddock reviewed (approved):
thanks for the added tests!

mmaietta commented:
😞 Looks like this needed to be rebased off latest `main` since the eslint PR went in (or vice versa). My apologies
